src/passes/MinimizeRecGroups.cpp

```diff
@@ -0,0 +1,702 @@
+
+  void run(Module* module) override {
+    // There are no recursion groups to minimize if GC is not enabled.
+    if (!module->features.hasGC()) {
+      return;
+    }
+
+    initBrandOptions();
+
+    types = ModuleUtils::getPrivateHeapTypes(*module);
+    for (auto type : ModuleUtils::collectHeapTypes(*module)) {
+      typeIndices.insert({type, typeIndices.size()});
+    }
+
+    // The number of types to optimize is an upper bound on the number of
+    // recursion groups we will emit.
+    groups.reserve(types.size());
+
+    // Compute the strongly connected components and ensure they form distinct
+    // recursion groups.
+    for (auto scc : TypeSCCs(types)) {
+      [[maybe_unused]] size_t index = equivalenceClasses.addSet();
+      assert(index == groups.size());
+      groups.emplace_back();
+
+      // The SCC is not necessarily topologically sorted to have the supertypes
+      // come first. Fix that.
+      std::vector<HeapType> sccTypes(scc.begin(), scc.end());
+      auto deps = createSubtypeGraph(sccTypes);
+      auto permutation = *TopologicalOrders(deps).begin();
+      groups.back().group.resize(sccTypes.size());
+      for (size_t i = 0; i < sccTypes.size(); ++i) {
+        groups.back().group[i] = sccTypes[permutation[i]];
+      }
+      updateShape(index);
+    }
+
+    rewriteTypes(*module);
+  }
+
+  void initBrandOptions() {
+    // Initialize the field options for brand types lazily here to avoid
+    // depending on global constructor ordering.
+    [[maybe_unused]] static bool fieldsInitialized = []() {
+      BrandTypeIterator::initFieldOptions();
+      return true;
+    }();
+  }
+
+  void updateShape(size_t group) {
+    auto [it, inserted] =
+      groupShapeIndices.insert({RecGroupShape(groups[group].group), group});
+    if (inserted) {
+      // This shape was unique. We're done.
+      return;
+    }
+
+    // We have a conflict. There are five possibilities:
+    //
+    //   1. We are trying to insert the next permutation of an existing
+    //      equivalence class and have found that...
+    //
+    //     A. The next permutation is an automorphism of some previous
+    //        permutation of the same equivalence class.
+    //
+    //     B. The next permutation is equivalent to some other existing group
+    //        shape already in a different nontrivial equivalent class.
+    //
+    //     C. The next permutation is equivalent to some other existing group
+    //        shape not yet included in a nontrivial equivalence class.
+    //
+    //   2. We are inserting a new group shape as of yet unaffiliated with a
+    //      nontrival equivalence class and have found that...
+    //
+    //     A. It is the same shape as some group in an existing nontrival
+    //        equivalence class.
+    //
+    //     B. It is the same shape as some other group also not yet affiliated
+    //        with a nontrival equivalence class, so we have a new nontrivial
+    //        equivalence class.
```

Comment:
Brief examples of each of these 5 cases might be useful, I think. But I don't feel strongly.

Reply:
I'm not sure it's possible to make examples for these brief, since you'd have to show the non-trivial existing rec group equivalence classes, etc.

Reply:
Makes sense. Maybe there is some mathematical shorthand that can be used as an analogy? If not, no worries, just a thought.